proposals/0000-hasfield-redesign.rst

```diff
@@ -0,0 +1,1193 @@
+.. sectnum::
+.. contents::
+
+
+This proposal recommends improvements to the design of the ``HasField``
+typeclass (differing from those planned under `#158
+<https://github.com/ghc-proposals/ghc-proposals/pull/158>`_).  In particular,
+the proposed design supports updates using a new ``SetField`` class, adds
+support for unlifted datatypes and fields, and specifies laws for the classes.
+
+In order to keep this proposal simple, it does not yet propose adding support
+for type-changing update, which is left for a future proposal.
+
+
+Motivation
+----------
+
+Following `proposal #6 <https://github.com/ghc-proposals/ghc-proposals/pull/6>`_,
+GHC 8.2 introduced a special built-in typeclass ``HasField`` in the
+``GHC.Records`` module, defined thus::
+
+  class HasField (x :: k) r a | x r -> a where
+    getField :: r -> a
+
+When the constraint solver sees a constraint of the form ``HasField "foo" T a``,
+where ``T`` is a concrete datatype and ``foo`` is a symbol corresponding to one
+of its fields, and this field is in scope, the constraint will be solved
+automatically with a dictionary derived from the record selector function for
+the field.
+
+This makes it possible to get a form of type-directed name resolution for field
+selection: given the expression ``getField @"foo" t``, the inferred type of
+``t`` can be used to determine which ``foo`` field is meant, even if there are
+multiple ``foo`` fields in scope and hence the expression ``foo t`` would be
+ambiguous.  (This arises in particular with the ``DuplicateRecordFields``
+extension, which has a somewhat ad hoc mechanism for disambiguating such
+expressions that has been removed in GHC 9.4, following `proposal #366
+<https://github.com/ghc-proposals/ghc-proposals/pull/366>`_.)
+
+GHC 9.2 includes support for using "record dot syntax" for selection with the
+``OverloadedRecordDot`` extension, e.g. ``t.foo`` can be used as syntactic sugar
+for ``getField @"foo" t``.  This is described in the accepted `proposal #282
+<https://github.com/ghc-proposals/ghc-proposals/pull/282>`_ (as modified by
+`proposal #405 <https://github.com/ghc-proposals/ghc-proposals/pull/405>`_).
+However, while the proposals describe both ``OverloadedRecordDot`` and another
+extension ``OverloadedRecordUpdate`` which allows type-based disambiguation of
+record update expressions, only the selection part is fully implemented so far.
+
+The accepted `proposal #158
+<https://github.com/ghc-proposals/ghc-proposals/pull/158>`_ plans to change the
+definition of ``HasField`` to support updates, which is necessary for the full
+implementation of the ``OverloadedRecordUpdate`` extension.
+An implementation of this proposal is available as `GHC merge request !3257
+<https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3257>`_, but has not yet
+been merged, because the compile-time performance cost of the selected
+implementation strategy is unacceptably high.  Such costs were not really
```

Comment:
You don't touch upon it in the proposal, but I assume that using a van Laarhoven lens type would solve the performance issue. It does pose other problems (in particular is not easy to make a such a type representation parametric), but presumably a word should be said somewhere in the _Alternatives_ section.